File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {

private: ListNode* res  =nullptr;

public:
    ListNode* reverseList(ListNode* head) {
        
        reverse(head);
        
        return res;
        /*
        ListNode* prev = nullptr;
        ListNode* current = head;
        ListNode* next = nullptr;
        
        while(current != nullptr){
            next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }
        
        return prev;
        */
    }
    
public:
    void reverse(ListNode* head){
        
        if(head == nullptr || head->next == nullptr){
            res = head;
            return;
        }
        
        reverse(head->next);
        
        head->next->next = head;
        head->next = nullptr;
        
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    private ListNode res = null;
    
    public ListNode reverseList(ListNode head) {
        
        /*
        // with iteration
        ListNode prev = null;
        ListNode current = head; 
        ListNode next = null;
        
        while(current != null){
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        
        
        return prev;
        */
        
        reverse(head);
        
        return res;
    }
    
    
    public void reverse(ListNode node){
        
        // with recursion
        
        if(node == null || node.next == null){
            res = node;
            return;
        }
        
        reverse(node.next);
        
        node.next.next = node;
        node.next = null;
    }
}